### DISCUSSION TITLE
wrapped native token

### DISCUSSION THREAD

**Discussion by weber77** [Cyfrin/security-and-auditing-full-course-s23#137] [Q&A]
When auditing the GMX protocol [there was a mention of re-wrapping a native token](https://youtu.be/pUWmJ86X_do?t=19099), in order for it to be accepted by a smart contract to prevent a DoS.

My questions:
1. Why do some smart contract not accept native tokens?
2. What is meant by re-wrapping a token?
3. How does re-wrapping make it acceptable?
4. How does preventing a position from been liquidated harm the protocol (Dumb one 😅)?

**Comment by EngrPips** [comment]
Hello @weber77 , Firstly let me start with mentioning that largely there is no dumb question as far as you are asking to understand clearly what is confusing to you. So now to your questions i would answer them in the order which you asked them 

**WHY DO SOME SMART CONTRACT NOT ACCEPT NATIVE TOKEN**
Smart contract Generally would not accept native token on ethereum blockchain except if they implement the `receive` or `fallback` function, so when a smart contract does not implement any of those two functions they wouldn't be able to receive ether except in an exceptional case where the ether is forced upon them with a `selfdestruct` function on some other contract [You should read about `selfdestruct` to know more]

**WHAT IS MEANT BY RE-WRAPPING A TOKEN**
In the context of the ethereum blockchain wrapping a token simply means to get the `ERC20` version of that token you want to wrap, honestly i am not sure how you mean `re-wrapping` as `wrapping` a token is what i know and understand properly.

**HOW DOES RE-WRAPPING MAKE IT ACCEPTABLE**
This is because receiving an `ERC20` token on the ethereum blockchain doesn't require a smart contract to implement a `receive` or `fallback` function

**HOW DOES PREVENTING A POSITION FROM BEEN LIQUIDATED HARM THE PROTOCOL**
This harm the protocol because not liquidating a position that needs to be liquidated would make it so the protocol have given out the amount of funds than they have in reserve.

imagine you borrowed $100 from a lending&borrowing platform by supplying them 2 ETH (assuming 1 ETH is $100), that means the protocol have $200 in reserve and have only given out $100, all of a sudden the value of ETH drop from (1ETH => $100) to (1ETH => $80) on seeing that the protocol tries to liquidate the position so that they end up with a balance of $160 (your 2ETH x the current price ($80)) but they was unable to do so and then the price of ETH further decline to (1ETH => $20) at this point the protocol borrowed you $100 but now all they have in reserve is $40 (your 2ETH x the current price ($20)). Now is the protocol running a profitable business or a business that would close down ?

**Comment by alymurtazamemon** [comment]
@weber77 Here are the answers;

1. Why do some smart contracts not accept native tokens?

Ans: Native tokens are just limited to a blockchain. For example, ethereum has its native token `ether`. These native tokens were created for gas purposes (so the blockchains can achieve Turing completeness) and cannot be traded/go on other blockchains. For that reason, developers created the ERC20 standard (here you can read about its motivation [link](https://eips.ethereum.org/EIPS/eip-20)) so they can be traded easily everywhere. It doesn't mean smart contracts cannot accept native tokens, if two contracts are within the same blockchain then they can use native tokens if they want, it all depends on their functionality.

2. What is meant by re-wrapping a token?

Answer: First you should understand wrapping. Wrapping is converting the actual assets such as BTC, or ETH to a wrapped token so they can perform the protocol's operations. We need this for two reasons;

1. On the DeFi protocols, when you deposit your assets, they give their wrapped tokens in return as a receipt, which you can use to claim back your actual assets.
2. To trade the native tokens of the blockchain, we need to wrap them to the ERC20 standard. For example, WETH. It is a wrapped version of `ether`. When you trade ether with another pair, the protocols first convert ethers to WETH and then trade them with another pair. You can also wrap your ether to WETH tokens using the WETH contract.

Re-wrapping is just the opposite of all this. When you want your actual assets back then you go to the protocol and provide them with their wrapped tokens and in return, they provide your assets back.

3. It is also covered in the 2.

4. How does preventing a position from being liquidated harm the protocol?

Ans: The protocols give an option to users to deposit their assets (lending) to get some other tokens in return (borrowing). Users want to do this for many reasons. Let's say Alice has 1 ETH (value $2000) but he doesn't want to sell it because he knows its value will go up soon, but he needs some money urgently. So he has the option to deposit his ETH to a DeFi platform and take some USDT in return for some additional fee which he needs to pay for this service.

However, protocols only allow borrowing up to 80% of the value of actual tokens (it depends on a token to token) so the protocol should always be in profit. Alice borrowed 1600 USDT worth of $1600 from a protocol and went to resolve his issues.

Suddenly, the ETH price decreases to $1700 which is very near to $1600 which Alice borrowed from the protocol. Now protocol wants its money back because if it doesn't do this then soon it will be in loss. So it will allow other members of the protocol to pay Alice's borrowed amount ($1600) and get his deposited ETH (worth $1700). Others will see a $100 profit in this situation then they will happily want to this trade. This way protocol will be safe from loss.

Now you can understand if anyone stops this process then soon protocol will face a loss.

**Comment by allwin199** [comment]
Learning a lot from the explanations.

Thanks :) @alymurtazamemon @EngrPips
